constraint-solver/src/solver/boolean_extractor.rs

```diff
@@ -1,71 +1,131 @@
+use std::{collections::HashMap, hash::Hash};
+
+use itertools::Itertools;
+
+use crate::{
+    constraint_system::ConstraintSystem,
+    grouped_expression::GroupedExpression,
+    indexed_constraint_system::apply_substitutions,
+    runtime_constant::{RuntimeConstant, Substitutable},
+    solver::VariableAssignment,
+};
+
+pub struct BooleanExtractor<T, V> {
+    /// If (expr, z) is in the map, it means that
+    /// we have transformed a constraint `left * right = 0` into
+    /// `right + z * offset = 0`, where `z` is a new boolean variable
+    /// and `expr = -right / offset = z`.
+    substitutions: HashMap<GroupedExpression<T, V>, V>,
+}
+
+impl<T, V> Default for BooleanExtractor<T, V> {
+    fn default() -> Self {
+        Self {
+            substitutions: HashMap::new(),
         }
-        _ => T::one(),
-    };
-
-    // `constr = 0` is equivalent to `left * right = 0`
-    let offset = &(left.clone() * &factor) - right;
-    // We only do the transformation if `offset` is known, because
-    // otherwise the constraint stays quadratic.
-    let offset = offset.try_to_known()?;
-    // We know that `offset + right = left` and thus
-    // `constr = 0` is equivalent to `right * (right + offset) = 0`
-    // which is equivalent to `right + z * offset = 0` for a new
-    // boolean variable `z`.
-
-    if offset.is_zero() {
-        // In this special case, we do not need a new variable.
-        Some(right.clone())
-    } else if (right.clone() * -T::one().field_div(offset))
-        .try_to_simple_unknown()
-        .is_some()
-    {
-        // In this case we don't gain anything because the new variable `z` will just
-        // be equivalent to the single variable in `right`.
-        return None;
-    } else {
-        let z = var_dispenser();
-
-        // We return `right + z * offset == 0`, which is equivalent to the original constraint.
-        Some(right + &(GroupedExpression::from_unknown_variable(z) * offset))
+    }
+}
+
+impl<T: RuntimeConstant + Hash, V: Ord + Clone + Hash + Eq> BooleanExtractor<T, V> {
+    /// Tries to simplify a quadratic constraint by transforming it into an affine
+    /// constraint that makes use of a new boolean variable.
+    /// NOTE: The boolean constraint is not part of the output.
+    ///
+    /// If the same simplification has been performed before, it will
+    /// return None (in particular, it will not request a new variable).
+    ///
+    /// For example `(a + b) * (a + b + 10) = 0` can be transformed into
+    /// `a + b + z * 10 = 0`, where `z` is a new boolean variable.
+    ///
+    /// @param constraint The quadratic constraint to transform.
+    /// @param var_dispenser A function that returns a new variable that is assumed to be boolean-constrained.
+    /// It will only be called if the transformation is performed.
+    pub fn try_extract_boolean(
+        &mut self,
+        constraint: &GroupedExpression<T, V>,
+        mut var_dispenser: impl FnMut() -> V,
```

Comment:
Why can't `var_dispenser` be a member of `self` here?

Reply:
The two components (linearizer and boolean extractor) share the variable dispenser, but they actually don't need to. Let me take a look, but I guess this is unrelated to this PR.

Reply:
Right, not blocking.